coffee/web_socket_channel.coffee

```diff
@@ -30,18 +31,41 @@ class palava.WebSocketChannel extends @EventEmitter
       @emit 'open', handshake
     @socket.onmessage = (msg) =>
       try
-        @emit 'message', JSON.parse(msg.data)
+        parsedMsg = JSON.parse(msg.data)
+        if parsedMsg == {event: "pong"}
+          @pongReceived = true
+          @contiguousPongsMissed = 0
+        else
+          @emit 'message', parsedMsg
       catch SyntaxError
         @emit 'error_invalid_json', msg
     @socket.onerror = (msg) =>
-      if @retries > 0
-        @retries -= 1
-        @setupWebsocket()
-      else
-        @emit 'error', msg
+      @retryWebsocketConnection()
     @socket.onclose = =>
+      clearInterval(@pingInterval)
       @emit 'close'
 
+  retryWebsocketConnection: =>
+    clearInterval(@pingInterval)
+    if @retries > 0
+      @retries -= 1
+      @setupWebsocket()
+      @startClientPings()
+    else
+      @emit 'error', msg
+
+  startClientPings: =>
+    @pongReceived = true
+    @contigousPongsMissed = 0
+    @pingInterval = setInterval( () =>
+      unless @pongReceived
+        @contiguousPongsMissed += 1
+        if @contiguousPongsMissed >= 6
+          @retryWebsocketConnection()
```

Comment:
Wouldn't this lose all the state attached to the websocket (user id changes, no longer in room) without telling the user of this object?

Reply:
> `pongReceived` could be removed turning `contiguousPongsMissed` into `outstandingPongs` serving the same purpose (after increasing the threshold by 1) decreasing code complexity.

You're right, thanks, I improved that!

Reply:
True, the retry should not happen there, so I now emit the error right away.

Reply:
(but it also wouldn't have been an error, because the retry does only happen when the connection was never established - @retries is set to 0 once the websocket connection is open)